logzio-sender/src/main/java/io/logz/sender/HttpsSyncSender.java

```diff
@@ -118,7 +84,38 @@ public void sendToLogzio(List<FormattedLogMessage> messages) throws LogzioServer
         }
     }
 
-    private void readBadRequest(HttpURLConnection conn) {
+    private int tryBackoff(int currentRetrySleep, int currTry, int responseCode, String responseMessage, IOException savedException) throws LogzioServerErrorException, InterruptedException {
+        if (currTry == configuration.getMaxRetriesAttempts()) {
+            if (savedException != null) {
+                reporter.error("Got IO exception on the last bulk try to logz.io", savedException);
+            }
+            // Giving up, something is broken on Logz.io side, we will try again later
+            throw new LogzioServerErrorException("Got HTTP " + responseCode + " code from logz.io, with message: " + responseMessage);
+        }
+
+        reporter.warning("Could not send log to logz.io, retry (" + currTry + "/" + configuration.getMaxRetriesAttempts() + ")");
+        reporter.warning("Sleeping for " + currentRetrySleep + " ms and will try again.");
+        Thread.sleep(currentRetrySleep);
+        return currentRetrySleep * 2;
+    }
+
+    private boolean shouldRetry(byte[] payload, int responseCode, String responseMessage, HttpURLConnection conn) {
```

Comment:
I'll say that again - I don't like the way this code is built. Next PR we must refactor how this logic works.